### PR TITLE
[BEAM-1092] Shade commonly used libraries (e.g. Guava) to avoid class conflicts

### PR DESCRIPTION
**Pull request by aviemzur**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by aviemzur** [comment]
R: @davorbonaci @dhalperi 
CC: @francesperry @jbonofre

**Comment by dhalperi** [comment]
We looked into this exact approach in the past, and rolled it back because `project.artifactId` is not a legal java package name (e.g., if it included `-`). Did you find a way to solve this?

**Comment by dhalperi** [comment]
(#744)

**Comment by aviemzur** [comment]
I tried build helper maven plugin and was able to create a property that replaces the `-` with `.` but it happens after it sets the configuration for the shade plugin (Tried running it in different phases to no avail).

**Comment by aviemzur** [comment]
I can replace the `-` with `.`

**Comment by aviemzur** [comment]
Retest this please.

**Comment by aviemzur** [comment]
@dhalperi Addressed your comment in my latest commit. PTAL?

File: sdks/java/core/pom.xml [hunk 1]
```diff
@@ -84,54 +84,6 @@
...
-                </filter>
-              </filters>
-              <relocations>
-                <!-- TODO: Once ready, change the following pattern to 'com' only, 
-                  exclude 'org.apache.beam.**', and remove the second relocation. -->
-                <relocation>
-                  <pattern>com.google.common</pattern>
-                  <excludes>
-                    <!-- com.google.common is too generic, need to exclude guava-testlib -->
-                    <exclude>com.google.common.**.testing.*</exclude>
```
Comment (dhalperi) [inline sdks/java/core/pom.xml]: this looks like it's being lost. Is this right?

File: pom.xml [hunk 2]
```diff
@@ -1457,6 +1525,14 @@
           </dependency>
         </dependencies>
       </plugin>
+      <plugin>
+        <groupId>org.codehaus.mojo</groupId>
+        <artifactId>build-helper-maven-plugin</artifactId>
+      </plugin>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
```
Comment (dhalperi) [inline pom.xml]: how do we ensure this only executes in the modules that use guava?

What happens in the modules that need to shade more than guava?

File: runners/direct-java/pom.xml [hunk 3]
```diff
@@ -86,55 +86,6 @@
...
-              </filters>
-              <relocations>
-                <!-- TODO: Once ready, change the following pattern to 'com'
-                     only, exclude 'org.apache.beam.**', and remove
-                     the second relocation. -->
-                <relocation>
-                  <pattern>com.google.common</pattern>
-                  <excludes>
-                    <!-- com.google.common is too generic, need to exclude guava-testlib -->
-                    <exclude>com.google.common.**.testing.*</exclude>
```
Comment (dhalperi) [inline runners/direct-java/pom.xml]: same q?

File: sdks/java/core/pom.xml (hunk 1 above)
Comment (aviemzur) [inline sdks/java/core/pom.xml]: After my change core test jar is not longer shaded (Or any test jar for that matter). I didn't see any module failing because of this.
If this is still needed - If anyone outside of Beam build cycle uses the core test jar - we can add shading for its test jar specifically in its pom.

File: pom.xml (hunk 2 above)
Comment (aviemzur) [inline pom.xml]: This will execute in every module, to ensure forward compatability, when new modules are introduced, they will not need to shade guava.
For modules that need to shade more than guava we should ask the question, should this be shaded in all modules that will use it? If so add it to the root pom, if not override shading for that specific module.

**Comment by dhalperi** [comment]
Re: testlib -- I did some GitHub blame/history diving and found JIRA https://issues.apache.org/jira/browse/BEAM-557 and https://github.com/apache/beam/pull/832 .

Maybe @swegner has more context about which specific ways the Guava testlib is a public dependency.

**Comment by dhalperi** [comment]
Re: general approach.

I see two improvements:

1. Providing a simple, uniform way to name shaded artifacts by establishing a central rewritten `artifactId`. This is succeeding where #744 failed!
2. Providing a centralized way to ensure that Guava is not on the public API surface of any module.

I see two downsides:

1. Extra executions for modules that don't depend on Guava. This will slow the build.
2. Modules that need to shade more than just Guava will have to override the shading config. They may actually get it wrong after that, so Guava might still be exposed. -- Thus, while this reduces the amount of config necessary in common cases, does guarantee solution.

I would probably:
* move the shading config back into individual modules. build perf + accuracy. But use the new `artifactId` trick to make the config more uniform.
* find a better way to actually verify that the final output artifacts don't expose Guava. (out of scope for this PR.) This is in-line with prior PR comments.

In otherwords, I would probably take improvement 1, but not 2. 1 is a clear win, but 2 has some drawbacks and doesn't quite guarantee success to justify those drawbacks IMO.

**Comment by dhalperi** [comment]
Another good opinion to get would be @davorbonaci .

File: sdks/java/core/pom.xml (hunk 1 above)
Comment (swegner) [inline sdks/java/core/pom.xml]: This was added in PR #832 because the rules only shade `com.google.guava:guava` and not `com.google.guava:guava-testlib`. If we relocate symbols from `guava-testlib` without shading them, we can get runtime `ClassNotFoundException`s on projects which depend on our artifacts.

IIRC, this issue did not manifest within the the beam project structure because each project had equivalent repackaging rules and could source the repackaged `guava-testlib` symbols from its own local dependencies. I believe I ran into the issue when creating my own project that didn't have shading and consumed the shaded beam artifacts.

I've been away from Beam development for some months so unfortunately I don't have recent context on whether this is still needed.

Comment (aviemzur) [inline sdks/java/core/pom.xml]: Thanks @swegner  for your comment!

All dependencies on `guava-testlib` in the project currently are in scope `test`.

I removed shading of test jars in modules which matched the shading rules that are in the root pom in my branch (They now do not shade their test jars)

The only test jar that remains with shading is the one of Dataflow runner, which still has this exclusion in its shade-plugin configuration: [google-cloud-dataflow-java/pom.xml](https://github.com/aviemzur/beam/blob/shade-guava-generically/runners/google-cloud-dataflow-java/pom.xml#L138)

Is this sufficient?

Comment (swegner) [inline sdks/java/core/pom.xml]: Sounds good to me; thanks for the thorough investigation!

(Please continue working with @dhalperi for the rest of the PR)

**Comment by aviemzur** [comment]
@dhalperi I understand the concerns you raise. I'll try and address them from my point of view.

> 1. Extra executions for modules that don't depend on Guava. This will slow the build.

Many of the modules already shade, so regarding slowing down of the build I’m not sure this concern is that pertinent if new way of shading is an improvement.

> 2. Modules that need to shade more than just Guava will have to override the shading config. They may 
> actually get it wrong after that, so Guava might still be exposed. -- Thus, while this reduces the amount of config necessary in common cases, does guarantee solution.

I agree that human error could cause Guava to be forgotten from shading configuration added to child modules, but is this different than today? Can this human error not happen now?

**Comment by davorbonaci** [comment]
I'm not totally sold on the premise that we should shade Guava in every single module. A couple of reasons:
* We now have ~30 modules. We'll probably have 50+ sometime this year. Guava is not that small in size -- bringing in so many copies is just wasteful for a simple call to, say, `Preconditions`. Minimization could mitigate this concern, but I'm yet-to-be-convinced it works well for libraries.
* We care that Guava is not leaked to our users -- this is totally legit. On the other hand, we don't have a strong reason to avoid Guava on an internal interface between, say, `runners-core` and `core-sdk`. It just doesn't matter to users. Of course, taking a stronger stance can be fine, but I'm a little worried it will bite us (for no real reason).
* Build time.

Separately, the amount of "magic" worries me too. This will work perfectly in Maven, but what happens when someone imports it into Eclipse, IntelliJ or elsewhere? At least in some IDEs, it will fail.

On the other hand, I'm really strongly in favor of one common configuration. There's literally zero chance that this will work if this configuration is offloaded to every single module. This one outweighs them all for me. (And, I was bitten by Guava 19 vs. 20 just today -- so, I know the pain.)

In summary, I'm leaning towards +1 on this. However, I'd be more comfortable if we were explicit in saying:
* how much the build time has gone up
* how much the total bundled jar, across the project, has gone up
* to investigate minimization (separately from this PR)
* that we'll make enforcement that Guava in not brought in transitively (which is currently broken at least in one module).

**Comment by aviemzur** [comment]
Build time (MacBook Pro, skipping tests)
-
**Before:** 02:13 min
**After:** 02:35 min

Jar size (Sum of all jars generated)
-
**Before:** 93.93 MB
**After:** 151.34 MB

**Comment by davorbonaci** [comment]
Build time -- less than I would have forecasted. SGTM.

Jar size -- more than I would have forecasted -- it is 50% more. I think minimization is in order.

**Comment by aviemzur** [comment]
* Rebased on top of master.
* Added ServiceResourceTransformer to the common shading configuration to match what was done in #2182 

Regarding file sizes:

Right now this change adds shading and repackaging of Guava to all modules which are dependent on Guava and did not configure shading and repackaging.

As we know, many of the modules need this since leaking Guava to a user dependent on them can cause issues.

If there are modules which do not need this, and we feel that their filesize has inflated without any reason, we can add the following to them to "opt-out" of the default shading configuration:
```xml
<plugin>
  <groupId>org.apache.maven.plugins</groupId>
  <artifactId>maven-surefire-plugin</artifactId>
  <configuration>
    <systemPropertyVariables>
      <beamUseDummyRunner>false</beamUseDummyRunner>
    </systemPropertyVariables>
  </configuration>
</plugin>
<plugin>
  <groupId>org.apache.maven.plugins</groupId>
  <artifactId>maven-shade-plugin</artifactId>
  <executions>
    <execution>
      <id>disable-shading</id>
      <phase>package</phase>
      <goals>
        <goal>shade</goal>
      </goals>
      <configuration>
        <artifactSet>
          <excludes>
            <exclude>*</exclude>
          </excludes>
        </artifactSet>
      </configuration>
    </execution>
  </executions>
</plugin>
```

Another way to tackle this would be to configure minimization by default.

I think we should start a minimization track separately from this change, as there are some issues with it:
1. There is a bug in shade-plugin where if minimization is declared on a project with pom packaging it fails. I have created a [ticket](https://issues.apache.org/jira/browse/MSHADE-253) and [PR](https://github.com/apache/maven-plugins/pull/107) with a fix for this but could take a while before it is available to us.
2. Minimization on all of our modules adds a significant amount of build time, which, as discussed, we wish to avoid. (Up from 02:30 mins for a build skipping tests on my laptop to over 5 mins)

**Comment by davorbonaci** [comment]
Btw, I notices the version of the shade plugin went down. Was this intentional? If yes, great. If not, perhaps push a separate PR to get it back up.
